presto-jdbc/src/main/java/com/facebook/presto/jdbc/ObjectCasts.java

```diff
@@ -0,0 +1,253 @@
+            throw invalidConversion(x, targetSqlType, e);
+        }
+        throw invalidConversion(x, targetSqlType);
+    }
+
+    public static double castToDouble(Object x, int targetSqlType)
+            throws SQLException
+    {
+        if (x instanceof Boolean) {
+            return (((Boolean) x) ? 1 : 0);
+        }
+        try {
+            if (x instanceof Number) {
+                return ((Number) x).doubleValue();
+            }
+            if (x instanceof String) {
+                return parseDouble((String) x);
+            }
+        }
+        catch (RuntimeException e) {
+            throw invalidConversion(x, targetSqlType, e);
+        }
+        throw invalidConversion(x, targetSqlType);
+    }
+
+    public static BigDecimal castToBigDecimal(Object x, int targetSqlType)
+            throws SQLException
+    {
+        if (x instanceof Boolean) {
+            return BigDecimal.valueOf(((Boolean) x) ? 1 : 0);
+        }
+        if (x instanceof BigInteger) {
+            return new BigDecimal((BigInteger) x);
+        }
+        try {
+            if ((x instanceof Byte) || (x instanceof Short) || (x instanceof Integer) || (x instanceof Long)) {
+                return BigDecimal.valueOf(((Number) x).longValue());
+            }
+            if ((x instanceof Float) || (x instanceof Double)) {
+                return BigDecimal.valueOf(((Number) x).doubleValue());
+            }
+            if (x instanceof String) {
+                return new BigDecimal((String) x);
+            }
+        }
+        catch (RuntimeException e) {
+            throw invalidConversion(x, targetSqlType, e);
+        }
+        throw invalidConversion(x, targetSqlType);
+    }
+
+    public static byte[] castToBinary(Object x, int targetSqlType)
+            throws SQLException
+    {
+        if (x instanceof byte[]) {
+            return (byte[]) x;
+        }
+        try {
+            if (x instanceof String) {
+                return x.toString().getBytes(UTF_8);
+            }
+        }
+        catch (RuntimeException e) {
+            throw invalidConversion(x, targetSqlType, e);
+        }
+        throw invalidConversion(x, targetSqlType);
+    }
+
+    public static Date castToDate(Object x, int targetSqlType)
+            throws SQLException
+    {
+        if (x instanceof Date) {
+            return (Date) x;
+        }
+        if (x instanceof java.util.Date) {
+            return new Date(((java.util.Date) x).getTime());
+        }
+        if (x instanceof String) {
+            throw new SQLException("Conversion from String to Date not yet supported");
+        }
+        throw invalidConversion(x, targetSqlType);
+    }
+
+    public static Time castToTime(Object x, int targetSqlType)
+            throws SQLException
+    {
+        if (x instanceof Time) {
+            return (Time) x;
+        }
+        if (x instanceof java.util.Date) {
+            return new Time(((java.util.Date) x).getTime());
```

Comment:
Table B-5 lists conversion from `java.util.Date` to `TIME`. Are you thinking of `java.util.Date` to `DATE`? That one is strange because `java.util.Date` has a time but `java.sql.Date` does not. Though both are in the table.

Reply:
Indeed, conversion `java.util.Date` → `Types.TIME` should be supported, while `java.sql.Date` → `Types.TIME` is not listed (disallowed?)
But... the second extends the first, so technically you should `if (x instanceof java.util.Date && !(x instanceof java.sql.Date))`